compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            suggested_inputs[input_idx] = Some(format!("{{{}}}", input_ty));
+                        }
+                        if let Compatibility::Incompatible(coerce_error, subtyping_error) =
+                            &compatibility_matrix[idx][idx]
+                        {
+                            let provided_ty = final_arg_types[input_idx].map(|ty| ty.0).unwrap();
+                            let cause = &self.misc(provided_args[input_idx].span);
+                            let trace = TypeTrace::types(cause, true, expected_ty, provided_ty);
+                            if let Some(e) = coerce_error {
+                                self.note_type_err(
+                                    &mut err,
+                                    &trace.cause,
+                                    None,
+                                    Some(trace.values),
+                                    e,
+                                    false,
+                                );
+                            }
+                            if let Some(e) = subtyping_error {
+                                self.note_type_err(
+                                    &mut err,
+                                    &trace.cause,
+                                    None,
+                                    Some(trace.values),
+                                    e,
+                                    false,
+                                );
+                            }
+                        }
+
+                        self.emit_coerce_suggestions(
+                            &mut err,
+                            &provided_args[input_idx],
+                            final_arg_types[input_idx].map(|ty| ty.0).unwrap(),
+                            final_arg_types[input_idx].map(|ty| ty.1).unwrap(),
+                            None,
+                            None,
+                        );
+
+                        eliminate_input(&mut compatibility_matrix, &mut input_indexes, idx);
+                        eliminate_arg(&mut compatibility_matrix, &mut arg_indexes, idx);
+                    }
+                    Some(Issue::Extra(idx)) => {
+                        // Eliminate the argument (without touching any inputs)
+                        let arg_idx = arg_indexes[idx];
+                        let arg_type = if let Some((_, ty)) = final_arg_types[arg_idx] {
+                            format!(" of type {}", ty)
+                        } else {
+                            "".into()
+                        };
+                        labels.push((
+                            provided_args[arg_idx].span,
+                            format!("argument{} unexpected", arg_type),
+                        ));
+                        suggestion_text = match suggestion_text {
+                            None => Some("remove the extra argument"),
+                            Some(_) => Some("did you mean"),
```

Comment:
Let's think about an alternative to "did you mean"

Reply:
Can we leave this as a followup?